cpp/bench/ann/src/common/dataset.hpp

```diff
@@ -56,8 +58,21 @@ struct dataset {
   std::optional<blob<IdxT>> ground_truth_set_;
   std::optional<blob<bitset_carrier_type>> filter_bitset_;
 
-  mutable bool base_set_accessed_  = false;
-  mutable bool query_set_accessed_ = false;
+  // Protects the lazy mutations of the blobs accessed by multiple threads
+  mutable std::mutex mutex_;
+  // The dim can be read either from the training set or from the query set.
+  // This cache variable is filled from either of the two sets loaded first.
+  mutable std::atomic<int> dim_ = -1;
+
+  // Before returning the passed result, cache the dim value from the passed blob.
+  template <typename R>
+  auto with_cached_dim_from(const blob<DataT>& blob, R&& r) const -> R
```

Comment:
In order to cache dim from blob, we do not need `r`. I would recommend hat this function only caches the dimension. The call site can return `r`.

Reply:
The point of forwarding `r` here is to shorten the code at the call site. Consider what we have now:
```
return with_cached_dim_from(query_set_, query_set_.data());
```
against what would be without forwarding:
```
auto* r = query_set_.data());
cache_dim_using(query_set_);
return r;
```
(I compute `r` first here, so that if the exception arises due to a missing file or something else, the reason of failure is more easily traced)
The helper is currently used eight times, so the result is 8 lines against 24.

I don't insist though :) and surely ok to change this if it would be easier to read for others.

Reply:
Thanks Artem for explanation. It was not clear to me that we need to evaluate the expression beforehand. Still I when I look at the code
```c++
return with_cached_dim_from(query_set_, query_set_.data());
```
I have no idea what this is returning. In contrast, the more verbose option is clear. Therefore I strongly prefer without forwarding. (Sorry for splitting hairs.)